crates/kubelet/src/plugin_watcher/mod.rs

```diff
@@ -0,0 +1,744 @@
+
+const SOCKET_EXTENSION: &str = "sock";
+
+/// An enum for capturing possible plugin types. This is purely for clarity and capturing this
+/// information is a compiled type as the information we get from gRPC is a string
+enum PluginType {
+    CSIPlugin,
+    DevicePlugin,
+}
+
+impl TryFrom<&str> for PluginType {
+    type Error = anyhow::Error;
+
+    fn try_from(value: &str) -> Result<Self, Self::Error> {
+        match value {
+            "CSIPlugin" => Ok(PluginType::CSIPlugin),
+            "DevicePlugin" => Ok(PluginType::DevicePlugin),
+            _ => Err(anyhow::anyhow!(
+                "Unknown plugin type {}. Allowed types are 'CSIPlugin' and 'DevicePlugin'",
+                value
+            )),
+        }
+    }
+}
+
+/// Internal storage structure for a plugin
+#[derive(Debug)]
+struct PluginEntry {
+    plugin_path: PathBuf,
+    endpoint: Option<PathBuf>,
+}
+
+/// An internal storage plugin registry that implements most the same functionality as the [plugin
+/// manager](https://github.com/kubernetes/kubernetes/tree/fd74333a971e2048b5fb2b692a9e043483d63fba/pkg/kubelet/pluginmanager)
+/// in kubelet
+pub struct PluginRegistrar {
+    plugins: RwLock<HashMap<String, PluginEntry>>,
+    plugin_dir: PathBuf,
+}
+
+impl Default for PluginRegistrar {
+    fn default() -> Self {
+        PluginRegistrar {
+            plugin_dir: PathBuf::from(DEFAULT_PLUGIN_PATH),
+            plugins: RwLock::new(HashMap::new()),
+        }
+    }
+}
+
+impl PluginRegistrar {
+    /// Returns a new plugin registrar configured with the given plugin directory path
+    pub fn new<P: AsRef<Path>>(plugin_dir: P) -> Self {
+        PluginRegistrar {
+            plugin_dir: PathBuf::from(plugin_dir.as_ref()),
+            ..Default::default()
+        }
+    }
+
+    /// Gets the endpoint for the given plugin name, returning `None` if it doesn't exist
+    pub async fn get(&self, plugin_name: &str) -> Option<PathBuf> {
+        let lock = self.plugins.read().await;
+        lock.get(plugin_name)
+            .map(|v| v.endpoint.as_ref().unwrap_or(&v.plugin_path).to_owned())
+    }
+
+    /// Starts the plugin registrar and runs all automatic plugin discovery and registration loops
+    pub async fn run(&self) -> anyhow::Result<()> {
+        // Create plugin directory if it doesn't exist
+        create_dir_all(&self.plugin_dir).await?;
+
+        // Walk the plugin dir before hand and process any currently existing sockets
+        let dir_entries: Vec<PathBuf> = read_dir(&self.plugin_dir)
+            .await?
+            .map(|res| res.map(|entry| entry.path()))
+            .collect::<Result<Vec<PathBuf>, _>>()
+            .await?;
+
+        // Manually assemble an event and call handle_event to reconfigure sockets properly on restart
+        self.handle_create(Event {
+            paths: dir_entries,
+            ..Default::default()
+        })
+        .await?;
+
+        let mut event_stream = FileSystemWatcher::new(&self.plugin_dir)?;
```

Comment:
Yes there definitely is a possible race condition here (as seen by my delay in the tests). Honestly not quite sure how to handle that situation here without rescanning the directory one more time before dropping into the loop

Reply:
Could you handle it by having the watcher raise an event for the initial set of files instead of only for changes?  Or do the built-in ones not support that behaviour?

Otherwise I guess you need to start the watcher, then do the scan, and handle the possibility of a duplicate (a file that is created immediately after the watcher starts, so gets a Create event, but is also picked up in the subsequent "initial state" scan).

Reply:
I don't think the built in ones support that behavior, definitely not without escaping down to OS specific code. But even then I don't think they do